### ISSUE TITLE
[BUG] screenshots from another test are attached on failure

### ISSUE THREAD

**Issue by mk654321** [microsoft/playwright#24287]
### System info
- Playwright Version: v1.36.1
- Operating System: linux
- Browser: [All]
- Other info:

### Source code

- [ ] I provided exact source code that allows reproducing the issue locally.

**Link to the GitHub repository with the repro**

or

**Config file**

```js
// playwright.config.ts
// @ts-check
const { defineConfig, devices } = require('@playwright/test');

/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
// require('dotenv').config();

/**
 * @see https://playwright.dev/docs/test-configuration
 */
module.exports = defineConfig({
  testDir: './tests',
  /* Run tests in files in parallel */
  fullyParallel: true,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: 'html',
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Base URL to use in actions like `await page.goto('/')`. */
    // baseURL: 'http://127.0.0.1:3000',

    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'on-first-retry',
    /* screenshots on failure */
    screenshot: 'only-on-failure',

    //video: 'on'
  },

  /* Configure projects for major browsers */
  projects: [

    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] },
    },

    {
      name: 'firefox',
      use: { ...devices['Desktop Firefox'] },

    },

    {
      name: 'webkit',
      use: { ...devices['Desktop Safari'] },
    },

    /* Test against mobile viewports. */
    // {
    //   name: 'Mobile Chrome',
    //   use: { ...devices['Pixel 5'] },
    // },
    // {
    //   name: 'Mobile Safari',
    //   use: { ...devices['iPhone 12'] },
    // },

    /* Test against branded browsers. */
    // {
    //   name: 'Microsoft Edge',
    //   use: { ...devices['Desktop Edge'], channel: 'msedge' },
    // },
    // {
    //   name: 'Google Chrome',
    //   use: { ..devices['Desktop Chrome'], channel: 'chrome' },
    // },
  ],

  /* Run your local dev server before starting the tests */
  // webServer: {
  //   command: 'npm run start',
  //   url: 'http://127.0.0.1:3000',
  //   reuseExistingServer: !process.env.CI,
  // },
});


```

**Test file (self-contained)**

test_1.spec.js
```js
// @ts-check
const { test, expect } = require('@playwright/test');

/** @type {import('@playwright/test').Page} */
let page;

test.beforeAll(async ({ browser }) => {
  page = await browser.newPage();
  await page.goto("https://bing.com");
});

test.afterAll(async () => {
  //await page.close();
});




test('Async test 1', async () => {
    await page.goto('https://playwright.dev/community/welcome');
    let heading = page.getByRole('heading', { name: 'Welcome' });
    await expect(heading).toBeVisible();

  });

test('Async test 2', async () => {
    await page.goto('https://playwright.dev/docs/api/class-playwright');
    let heading = page.getByRole('heading', { name: 'Playwright Library' });
    await expect(heading).toBeVisible();
});

test('Steps are executed synchronously', async () => {
    await test.step('Step 1', async () => {
        await page.goto('https://playwright.dev/');
    });

    await test.step('Step 2', async () => {
        await page.getByText('Get Started').click();
    });

    await test.step('Step 3', async () => {
        let heading = page.getByRole('heading', { name: 'Installation' });
        await expect(heading).toBeVisible();
    });

});


test('Async test 3', async ({playwright}) => {


      let html = `<html>
      <head>
      <title>tets</title>
      <script>
      function loop() {
        document.getElementById('one').innerText = 'start';
        for(let i=0;i<500000;i++) {
          setTimeout(() => {
              document.getElementById('one').innerText = i +' '+ Date.now() +' '+ crypto.randomUUID();
          }, 50);



        }
      }
      </script>
      </head>
      <body onload="loop()">
      <h1 id='one'>pass</h1>

      </body>
      </html>`;
      await page.setContent(html);
      let heading = page.getByRole('heading');
      await expect(heading).toHaveText('fail');




    //await test.step('', async () => {
    //});

  });
```

test_2.spec.js
```js
// @ts-check
const { test, expect } = require('@playwright/test');

/** @type {import('@playwright/test').Page} */
let page;

test.beforeAll(async ({ browser }) => {

  page = await browser.newPage();
  await page.goto("https://example.com");

});

test.afterAll(async () => {
  //await page.close();
});




test('Async test 1', async () => {
    await page.goto('https://example.com/community/welcome');
    let heading = page.getByRole('heading', { name: 'Welcome' });
    await expect(heading).toBeVisible();

  });

test('Async test 2', async () => {
    await page.goto('https://example.com/docs/api/class-playwright');
    let heading = page.getByRole('heading', { name: 'Playwright Library' });
    await expect(heading).toBeVisible();
});

test('Steps are executed synchronously', async () => {
    await test.step('Step 1', async () => {
        await page.goto('https://example.com/');
    });

    await test.step('Step 2', async () => {
        await page.getByText('Get Started').click();
    });

    await test.step('Step 3', async () => {
        let heading = page.getByRole('heading', { name: 'Installation' });
        await expect(heading).toBeVisible();
    });

});


test('Async test 3', async ({playwright}) => {

      let html = `<html>
      <head>
      <title>tets</title>
      <script>
      function loop() {
        document.getElementById('one').innerText = 'start';
        for(let i=0;i<500000;i++) {
          setTimeout(() => {
              document.getElementById('one').innerText = i +' '+ Date.now() +' '+ crypto.randomUUID();
          }, 50);



        }
      }
      </script>
      </head>
      <body onload="loop()">
      <h1 id='one'>pass</h1>

      </body>
      </html>`;
      await page.setContent(html);
      let heading = page.getByRole('heading');
      await expect(heading).toHaveText('fail');




    //await test.step('', async () => {
    //});

  });
```

**Steps**
- Run the tests
- Review test_2.spec.js failures 

**Expected**

With the exception of "Async test 3" Only screenshots with example domain should be attached.

**Actual**

Usecase consists of config set to grab screenshots on failure and two test files of which one is failing all the tests. Even if tests are fully isolated , some of the screenshots show playwright.dev site instead of example.com 

https://playwright.dev/docs/browser-contexts reads:
Isolation
Tests written with Playwright execute in isolated clean-slate environments called browser contexts. This isolation model improves reproducibility and prevents cascading test failures.

What is Test Isolation?
Test Isolation is when each test is completely isolated from another test.

**Comment by mk654321** [comment]
Also I am not sure why but there are multiple screenshots per each fail - should this be reported as a separate issue?

**Comment by aslushnikov** [comment]
@mk654321 for the tests to be isolated between each other, you should actually use the [`page` fixture](https://playwright.dev/docs/test-fixtures#built-in-fixtures). 

I re-wrote your tests with the `page` fixture, now everything works as expected:

```ts
// test_1.spec.js
// @ts-check
const { test, expect } = require('@playwright/test');

test('Async test 1', async ({ page }) => {
  await page.goto('https://playwright.dev/community/welcome');
  let heading = page.getByRole('heading', { name: 'Welcome' });
  await expect(heading).toBeVisible();
});

test('Async test 2', async ({ page }) => {
  await page.goto('https://playwright.dev/docs/api/class-playwright');
  let heading = page.getByRole('heading', { name: 'Playwright Library' });
  await expect(heading).toBeVisible();
});

test('Steps are executed synchronously', async ({ page }) => {
  await test.step('Step 1', async () => {
    await page.goto('https://playwright.dev/');
  });

  await test.step('Step 2', async () => {
    await page.getByText('Get Started').click();
  });

  await test.step('Step 3', async () => {
    let heading = page.getByRole('heading', { name: 'Installation' });
    await expect(heading).toBeVisible();
  });
});

test('Async test 3', async ({ page }) => {
  let html = `<html>
  <head>
  <title>tets</title>
  <script>
  function loop() {
    document.getElementById('one').innerText = 'start';
    for(let i=0;i<500000;i++) {
      setTimeout(() => {
          document.getElementById('one').innerText = i +' '+ Date.now() +' '+ crypto.randomUUID();
      }, 50);
    }
  }
  </script>
  </head>
  <body onload="loop()">
  <h1 id='one'>pass</h1>

  </body>
  </html>`;
  await page.setContent(html);
  let heading = page.getByRole('heading');
  await expect(heading).toHaveText('fail');
  //await test.step('', async () => {
  //});
});
```

```ts
// test_2.spec.js
// @ts-check
const { test, expect } = require('@playwright/test');

test('Async test 1', async ({ page }) => {
  await page.goto('https://example.com/community/welcome');
  let heading = page.getByRole('heading', { name: 'Welcome' });
  await expect(heading).toBeVisible();
});

test('Async test 2', async ({ page }) => {
  await page.goto('https://example.com/docs/api/class-playwright');
  let heading = page.getByRole('heading', { name: 'Playwright Library' });
  await expect(heading).toBeVisible();
});

test('Steps are executed synchronously', async ({ page }) => {
  await test.step('Step 1', async () => {
    await page.goto('https://example.com/');
  });

  await test.step('Step 2', async () => {
    await page.getByText('Get Started').click();
  });

  await test.step('Step 3', async () => {
    let heading = page.getByRole('heading', { name: 'Installation' });
    await expect(heading).toBeVisible();
  });
});


test('Async test 3', async ({ page }) => {
  let html = `<html>
  <head>
  <title>tets</title>
  <script>
  function loop() {
    document.getElementById('one').innerText = 'start';
    for(let i=0;i<500000;i++) {
      setTimeout(() => {
          document.getElementById('one').innerText = i +' '+ Date.now() +' '+ crypto.randomUUID();
      }, 50);
    }
  }
  </script>
  </head>
  <body onload="loop()">
  <h1 id='one'>pass</h1>

  </body>
  </html>`;
  await page.setContent(html);
  let heading = page.getByRole('heading');
  await expect(heading).toHaveText('fail');
  //await test.step('', async () => {
  //});
});
```

Hope this helps!

**Comment by mk654321** [comment]
@aslushnikov thank you very much for your reply.
I am confused, mainly because I got screenshots from tests from another file and I thought that it shouldn't really matter if I use global page -  there should be full isolation between testfiles. If there isn't I have to take care of cleaning inbetween, this sounds to be in the contrary to
https://playwright.dev/docs/browser-contexts#two-ways-of-test-isolation
```
There are two different strategies when it comes to Test Isolation: start from scratch or cleanup in between. The problem with cleaning up in between tests is that it can be easy to forget to clean up and some things are impossible to clean up such as "visited links".
```

Also what if I want to authenticate as user 1 in test_1.spec.js and as user 2 in test_2.spec.js? (say I have 25 users, in 25 test files that I want to test, and being able to do that in parallel speeds testing up)
I thought of doing it by having global `page` within a test file (hoping they do not overlap between files) and authenticating inside `test.beforeAll` then running some tests.

**Comment by loknathloku41** [comment]
@aslushnikov - Same issue happening for us too..!! we are running a suite of 32 tests, where we have 5 failures. when i looked into one of the failed test it does have screenshots from other 4 failed tests. and sometimes we are seeing this issue as well
**Error: ENOENT: no such file or directory, open 'C:\GUI\Work\20230731134904-1281672-1422\test-results\Case_Tests-CM_Case_User_Tests-CMC-6ed0e-user-DetailID-Link-Test-b0eed0\test-finished-12.png'**

in our framework, we defined the page once per .spec file in test.beforeAll(). also using below configuration in playwright.config.js

browserName: 'chromium',
    headless: isHeadless,
    screenshot: 'on',
    trace: {
      mode: 'on',
      screenshots: true,
      sources: true,
      snapshots: false
    }

Please advise on this. Thank you
